### PR TITLE
release-1.6: Backports for Julia 1.6.8

### LINKED ISSUE #42245: jl_setjmp is not exported on Windows, making it impossible to use the JL_TRY/JL_CATCH macros when embedding Julia
On Windows `jl_setjmp` is implemented in assembly [here (64 bit)](https://github.com/JuliaLang/julia/blob/master/src/support/_setjmp.win64.S) and [here (32 bit)](https://github.com/JuliaLang/julia/blob/master/src/support/_setjmp.win32.S). On Linux, `jl_setjmp` is an "alias" to `sigsetjmp`. The result is that if the `JL_TRY`/`JL_CATCH` macros are used when embedding Julia on Linux they work successfully because `suigsetjmp` can be found in libc, but when embedding on Windows instead it fails because `jl_setjmp` is not exported.

The easiest way to reproduce this error is to wrap the call to `jl_eval_string` in the example found [here](https://docs.julialang.org/en/v1/manual/embedding/) with `JL_TRY`/`JL_CATCH`. I've used MINGW to compile the code, the instructions for MSVC didn't work but I haven't investigated this further because I have very little experience with developing on Windows and Visual Studio.

### LINKED ISSUE #43921: Product iterator is not zipped properly
When zipping with a product iterator with no arguments (which produces only one output, the empty tuple), it seems the check for zip completion does not handle this case:
```
julia> for (z,) in zip(Iterators.product())
       @show z
       end
z = ()
ERROR: ArgumentError: tuple must be non-empty
Stacktrace:
 [1] first(#unused#::Tuple{})
   @ Base ./tuple.jl:140
 [2] _pisdone
   @ ./iterators.jl:1005 [inlined]
 [3] isdone
   @ ./iterators.jl:1012 [inlined]
 [4] _zip_isdone(is::Tuple{Base.Iterators.ProductIterator{Tuple{}}}, ss::Tuple{Tuple{Bool}})
   @ Base.Iterators ./iterators.jl:379
 [5] _zip_iterate_all
   @ ./iterators.jl:348 [inlined]
 [6] iterate
   @ ./iterators.jl:341 [inlined]
 [7] top-level scope
   @ ./REPL[79]:2
```

`versioninfo()` output:
```
Julia Version 1.7.1
Commit ac5cc99908 (2021-12-22 19:35 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i7-8565U CPU @ 1.80GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-12.0.1 (ORCJIT, skylake)
Environment:
  JULIA_NUM_THREADS = 4
  JULIA_EDITOR = vim
```
EDIT: `zip([()])` works fine though:
```
julia> for (z,) in zip([()])
       @show z
       end
z = ()

julia> 
```

### LINKED ISSUE #45825: BitArray "constructor" incorrectly assumes 1-based indexing, leading to nondeterministic incorrect results due to out-of-bounds memory access
For example, in the following code, a `BitArray` that is constructed from an input array of values that are all `false` is found to contain some values that are `true`:

```jl
julia> using OffsetArrays

julia> all_false = falses(-1000:1000);

julia> any(==(true), all_false)
false

julia> all_false_bitarray = BitArray(all_false);

julia> any(==(true), all_false_bitarray)
true
```

This is due to the code [here](https://github.com/JuliaLang/julia/blob/2168230c210d620786311d91528a155582f78bc8/base/bitarray.jl#L510-L514), which applies `@inbounds` to a loop before proceeding to index the user-provided array with indices that start from one.

https://github.com/JuliaLang/julia/blob/2168230c210d620786311d91528a155582f78bc8/base/bitarray.jl#L510-L514

### LINKED ISSUE #45903: Indexed assignment with logical indices: subarray.jl error
Hi, 

I have run into the following error:
I have a BitArray whose entries I want to change based on another array

```julia
A = BitArray(rand(0:1, 3,3))
pos = randn(3,3)
A[pos .< 0]  .= 0
```
This yields the error message:

> ERROR: ArgumentError: number of indices (1) must match the parent dimensionality (2)
> Stacktrace:
>  [1] check_parent_index_match(parent::BitMatrix, #unused#::Tuple{Bool})
>    @ Base ./subarray.jl:43
>  [2] check_parent_index_match(parent::BitMatrix, indices::Tuple{Vector{Int64}})
>    @ Base ./subarray.jl:41
>  [3] SubArray
>    @ ./subarray.jl:21 [inlined]
>  [4] SubArray
>    @ ./subarray.jl:32 [inlined]
>  [5] SubArray
>    @ ./subarray.jl:28 [inlined]
>  [6] materialize!(B::Base.Broadcast.BitMaskedBitArray{2, 2}, bc::Base.Broadcast.Broadcasted{Base.Broadcast.DefaultArrayStyle{0}, Nothing, typeof(identity), Tuple{Int64}})
>    @ Base.Broadcast ./broadcast.jl:1174
>  [7] top-level scope
>    @ REPL[10]:1

The indexing works without an error, the error only appears when I try to assign the new value. 

Also, using Cartesian indices instead produces the expected result:

```julia
A = BitArray(rand(0:1, 3,3))
pos = randn(3,3)
A[findall(pos .< 0)]  .= 0
```

> versioninfo()
> Julia Version 1.7.2
> Commit bf53498635 (2022-02-06 15:21 UTC)
> Platform Info:
>   OS: macOS (arm64-apple-darwin21.2.0)
>   CPU: Apple M1 Pro
>   WORD_SIZE: 64
>   LIBM: libopenlibm
>   LLVM: libLLVM-12.0.1 (ORCJIT, cyclone)
> Environment:
>   JULIA_EDITOR = code
>   JULIA_NUM_THREADS =

### PR DESCRIPTION
**Pull request by KristofferC**
Backported PRs:
- [x] #42574 <!-- Fix sparse constructor when `Tridiagonal`/`SymTridiagonal` are empty -->
- [x] #44827 <!-- Fix implicit binding import for aliased bindings -->
- [x] #45552 <!-- `Base.stale_cachefile`: allow `ftime_req` to be greater than `ftime` by up to one microsecond (to compensate for Windows tar) -->
- [x] #45871 <!-- Fix integer overflow in `reverse!` -->
- [x] #45904 <!-- fix #45903, in place broadcast into a bit-masked bitmatrix -->
- [x] #45896 <!-- union-types: use insertion (stable) sort instead of qsort -->
- [x] #45605 <!-- Complete size checks in `BLAS.[sy/he]mm!` -->
- [x] #46040 <!-- fix convert call in Artifacts.jl -->
- [x] #46053 <!-- Allow `BUILDKITE_BRANCH` to provide branch name -->
- [x] #46163 <!-- fix rem2pi for non-finite arguments -->
- [x] #46161 <!-- `copyto!` fix for `BitArray`/`AbstractArray` -->
- [x] #46341 <!-- `leq` for reals falls back to `le` and `eq` -->
- [x] #46388 <!-- Handle `PhiNode` with `edge==0` -->
- [x] #45578 <!-- Fix union!(s::BitSet, r::AbstractUnitRange{<:Integer}) when two ranges do not overlap. -->
- [x] #46516 <!-- Use Documenter 0.27.23 -->
- [x] #46882 <!-- Set `intersection = 1` during `intersect_sub_datatype` -->
- [x] #46914 <!-- Don't crash on variable sized GC allocations -->
- [x] #45646 <!-- Replace isfinite check in ranges with lo ≤ x ≤ hi -->
- [x] #46938 <!-- bugfix: fld1 order of ops -->
- [x] #47114 <!-- Check sizes in 3-arg diagonal (dot-)product -->
- [x] #47058 <!-- make REPL completions robust against `expanduser` throwing -->
- [x] #46998 <!-- Better Libdl.dlopen error when using non-standard extension -->
- [x] #47299 <!-- Fix GC assertion on array of derived pointers -->
- [x] #47419 <!-- cgmemmgr,macos: add missing shared_map_lock initialization -->
- [x] #43947 <!-- Fix isdone for empty product iterators, fixes #43921 -->
- [x] #47433 <!-- CircleCI caching: allow ftime to be ceil(ftime_req) in Base.stale_cachefile -->
- [x] #47482 <!-- Removed attributes from arguments to gc_preserve_begin -->
- [x] #47352 <!-- Fix parsing of gfortran version -->
- [x] #47637 <!-- Fix REPL keybinding CTRL-Q for stdlib methods -->
- [x] #47739 <!-- Fix generator-invocation legality check for varargs generators -->
- [x] #47746 <!-- follow up #47739 -->
- [x] #47719 <!-- fix unescaping in `global` expressions -->
- [x] #47803 <!-- Set `OPENBLAS_NUM_THREADS=1` on local Distributed workers -->
- [x] #47782 <!-- Generalize Bool parse method to AbstractString -->
- [x] #48176 <!-- deps: fix broken `p7zip` url -->
- [x] #49096 <!-- doc/base: remove outdata const -->
- [x] #49124 <!-- Mark llvm::Any::TypeId as global in julia.expmap -->
- [x] #49212 <!-- Make :open ccall use variadic cconv -->
- [x] #50186 <!-- Remove xsaves from Zen cpuspec -->
- [x] #51254 <!-- Ryu: make sure adding zeros does not overwrite trailing dot -->

Need manual backport:
- [ ] #42112 <!-- inference: bail from const-prop if const-prop fails -->
- [ ] #42170 <!-- coverage: ensure topline is recorded during inlining -->
- [ ] #42267 <!-- fix #42245, export `jl_setjmp` on windows -->
- [ ] #44942 <!-- Properly set Vboxed field during union codegen -->
- [ ] #45155 <!-- codegen: add handling for undefined phinode values -->
- [ ] #45835 <!-- fix #45825, BitArray methods assuming 1-indexing of AbstractArray -->
- [ ] #46368 <!-- Properly declare alloca alignment -->
- [ ] #46473 <!-- Fix 3-arg `dot` for 1x1 structured matrices -->
- [ ] #46597 <!-- LibGit2: correct regex for credential helpers -->
- [ ] #47076 <!-- Fix alloc opt on unknown offset -->
- [ ] #47444 <!-- [CompilerSupportLibraries_jll] Update to new build -->
- [ ] #47499 <!-- syntax error with anonfn inside `elseif` and short-circuit op -->
- [ ] #47580 <!-- ensure bindings handle write barriers for ty and globalref -->
- [ ] #47986 <!-- Restore libgcc_s symlinking in !macOS -->
- [ ] #49357 <!-- Fix unclosed code fence in src/manual/methods.md -->
- [ ] #49636 <!-- Define a method for `hash(::Type, ::UInt)` -->
- [ ] #49909 <!-- [MPFR] Fix printing of `BigFloat` -->

Contains multiple commits, manual intervention needed:
- [ ] #42172 <!-- fix collect on stateful iterators -->
- [ ] #42837 <!-- fixed Float16 from Float64 and BigFloat -->
- [ ] #44476 <!-- process: ensure uvfinalize and _uv_close_cb are synchronized -->
- [ ] #44733 <!-- [Random] Document `default_rng` and remove references to `GLOBAL_RNG` -->
- [ ] #43510 <!-- Allow pushing to vector after 3-arg ldiv! -->
- [ ] #45009 <!-- Fix bug in `pinv` -->
- [ ] #45649 <!-- Emit aliases to FP16 conversion routines -->
- [ ] #46350 <!-- Fix intersect error reported in #41096 -->
- [ ] #47783 <!-- Prioritize build_dir for generated headers -->
- [ ] #47940 <!-- [PCRE2_jll] Upgrade to 10.42 -->
- [ ] #48562 <!-- Change uses of int to size_t in MurmurHash -->
- [ ] #50076 <!-- fix `hash(::BigInt)` on 32 bit systems -->
- [ ] #50267 <!-- `SuiteSparse_jll`: only`dlopen` the libraries if `Base.USE_GPL_LIBS` is true -->
- [ ] #50552 <!-- read(io, Char): fix read with too many leading ones -->
- [ ] #50989 <!-- fix incorrect results in `expm1(::Union{Float16, Float32})` -->

Non-merged PRs with backport label:
- [ ] #51667 <!-- [LibCURL_jll] Upgrade to v8.4.0 -->
- [ ] #51371 <!-- Workaround to improve `@debug` performance on Windows -->
- [ ] #49573 <!-- Implement jl_cpu_pause on PPC64 -->
- [ ] #42715 <!-- add a missing `transpose` in `LinearAlgebra.copy_transpose!` to handle matmul of isbits array-elements with non-identity transpose -->
- [ ] #41228 <!-- fix some issues with equality of factorizations -->
- [ ] #41041 <!-- fix command literals with trailing backslashes -->

### CODE REVIEW

**Comment by KristofferC** [comment]
@nanosoldier runtests(ALL, vs = ":release-1.6", configuration = (buildflags=["LLVM_ASSERTIONS=1", "FORCE_ASSERTIONS=1"],), vs_configuration=(buildflags=["LLVM_ASSERTIONS=1", "FORCE_ASSERTIONS=1"],))

**Comment by KristofferC** [comment]
@nanosoldier `runtests(ALL, vs = ":release-1.6", configuration = (buildflags=["LLVM_ASSERTIONS=1", "FORCE_ASSERTIONS=1"],), vs_configuration=(buildflags=["LLVM_ASSERTIONS=1", "FORCE_ASSERTIONS=1"],))`

**Comment by KristofferC** [comment]
@nanosoldier `runbenchmarks(ALL, vs = ":release-1.6")`

**Comment by nanosoldier** [comment]
[Your job](https://github.com/JuliaLang/julia/pull/46116#issuecomment-1361311461) failed. cc @maleadt

**Comment by KristofferC** [comment]
@nanosoldier `runtests(ALL, vs = ":release-1.6")`

**Comment by nanosoldier** [comment]
[Your job](https://github.com/JuliaLang/julia/pull/46116#issuecomment-1361331496) failed. cc @maleadt

**Comment by maleadt** [comment]
The build failed because gfortran was removed from the package rootfs, https://github.com/JuliaCI/rootfs-images/, resulting in broken triple detection:

```
Dec 21 08:58:14 amdci8 bash[65957]:       From worker 2:        │ curl: (22) The requested URL returned error: 404
Dec 21 08:58:14 amdci8 bash[65957]:       From worker 2:        │ make[2]: *** [/source/deps/mbedtls.mk:126: /source/deps/srccache/MbedTLS.v2.24.0+4..tar.gz] Error 22
Dec 21 08:58:14 amdci8 bash[65957]:       From worker 2:        │ make[2]: *** [/source/deps/pcre.mk:60: /source/deps/srccache/PCRE2.v10.40.0+0..tar.gz] Error 22
Dec 21 08:58:14 amdci8 bash[65957]: [152B blob data]
Dec 21 08:58:14 amdci8 bash[65957]:       From worker 2:        │ curl: (22) The requested URL returned error: 404
Dec 21 08:58:14 amdci8 bash[65957]:       From worker 2:        │ make[2]: *** [/source/deps/p7zip.mk:49: /source/deps/srccache/p7zip.v17.4.0+0..tar.gz] Error 22
Dec 21 08:58:14 amdci8 bash[65957]: [111B blob data]
Dec 21 08:58:14 amdci8 bash[65957]:       From worker 2:        │ curl: (22) The requested URL returned error: 404
Dec 21 08:58:14 amdci8 bash[65957]: [111B blob data]
Dec 21 08:58:14 amdci8 bash[65957]:       From worker 2:        │ curl: (22) The requested URL returned error: 404
Dec 21 08:58:14 amdci8 bash[65957]: [111B blob data]
Dec 21 08:58:14 amdci8 bash[65957]:       From worker 2:        │ curl: (22) The requested URL returned error: 404
Dec 21 08:58:14 amdci8 bash[65957]:       From worker 2:        │ make[2]: *** [/source/deps/mpfr.mk:80: /source/deps/srccache/MPFR.v4.1.1+0..tar.gz] Error 22
Dec 21 08:58:14 amdci8 bash[65957]:       From worker 2:        │ make[2]: *** [/source/deps/nghttp2.mk:56: /source/deps/srccache/nghttp2.v1.41.0+0..tar.gz] Error 22
Dec 21 08:58:14 amdci8 bash[65957]:       From worker 2:        │   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
Dec 21 08:58:14 amdci8 bash[65957]:       From worker 2:        │                                  Dload  Upload   Total   Spent    Left  Speed
Dec 21 08:58:14 amdci8 bash[65957]: [269B blob data]
Dec 21 08:58:14 amdci8 bash[65957]:       From worker 2:        │ curl: (22) The requested URL returned error: 404
```

Notice the wrong filenames, `MPFR.v4.1.1+0..tar.gz`. 

Maybe we can force this using a build flag; let's try:

@nanosoldier `runtests(ALL, vs = ":release-1.6", vs_configuration = (buildflags=["BB_TRIPLET_LIBGFORTRAN_CXXABI=x86_64-linux-gnu-libgfortran5-cxx11"],))`

If that doesn't work I'll deploy a workaround.

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/46116#issuecomment-1361311688) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/8ccf74a_vs_4688e54/report.md).

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/46116#issuecomment-1361349307) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/4cb3ce7_vs_4688e54/report.html).

File: src/intrinsics.cpp [hunk 1]
```diff
@@ -1474,8 +1474,6 @@ static inline uint16_t float_to_half(float param)
     return h;
 }
 
-#if !defined(_OS_DARWIN_)   // xcode already links compiler-rt
-
```
Comment (staticfloat) [inline src/intrinsics.cpp:1478]: @KristofferC I'm not sure if you want to merge this commit into your backport or not; it's up to you.

**Comment by staticfloat** [comment]
The `analyzegc` failure looks real: https://buildkite.com/julialang/julia-release-1-dot-6/builds/228#01853669-24f9-461b-8078-efeb7ed89c35/814-1059

`git blame` shows that @vtjnash is the last person to move around the `jl_lock_profile()` code, and it's a bit hard for me to tell; should this function be considered a safepoint or not?

**Comment by vtjnash** [comment]
I don't know about v1.6, but on master, it is annotated
```
src/julia_internal.h:JL_DLLEXPORT void jl_lock_profile(void) JL_NOTSAFEPOINT;
```

**Comment by staticfloat** [comment]
The windows build now makes it all the way to system image emission, and then [dies with an access violation in `_ZN12_GLOBAL__N_119WinCOFFObjectWriter11writeObjectERN4llvm11MCAssemblerERKNS1_11MCAsmLayoutE`](https://buildkite.com/julialang/julia-release-1-dot-6/builds/232#01853757-0de6-435b-910a-679a08ad4b5d/567-2233).  I can't say I've seen this before.

**Comment by staticfloat** [comment]
> I don't know about v1.6, but on master, it is annotated
> ```
> src/julia_internal.h:JL_DLLEXPORT void jl_lock_profile(void) JL_NOTSAFEPOINT;
> ```

Yes, that was [added by you](https://github.com/JuliaLang/julia/commit/f4d13cf791169a4afa503a31fb46f5a291645324#diff-7b3edb5362e7b70b75aff15c04b492aa3c74b883fca8eeee74d7bd839f81e4b0R139).  It looks like on standard `release-1.6`, the analyzegc pass fails, so I guess we can just ignore that.

**Comment by staticfloat** [comment]
The `llvmpasses` issue is due to some problem in our LLVM builds the assert build does not contain a needed symbol:

```
$ nm ./libLLVM.v11.0.1+3.x86_64-linux-gnu-cxx11/lib/libLLVM-11jl.so | grep DisableABIBreakingChecks
0000000003e9ec30 B _ZN4llvm24DisableABIBreakingChecksE
$ nm ./libLLVM_assert.v11.0.1+3.x86_64-linux-gnu-cxx11/lib/libLLVM-11jl.so | grep DisableABIBreakingChecks
$
```

So this may just be a problem with our earlier LLVM builds.  Once again, I think this is just something we didn't run much on 1.6.  I'm just going to disable `llvmpasses` and `analyzegc` on 1.6 for now.

**Comment by staticfloat** [comment]
> The windows build now makes it all the way to system image emission, and then [dies with an access violation in _ZN12_GLOBAL__N_119WinCOFFObjectWriter11writeObjectERN4llvm11MCAssemblerERKNS1_11MCAsmLayoutE](https://buildkite.com/julialang/julia-release-1-dot-6/builds/232#01853757-0de6-435b-910a-679a08ad4b5d/567-2233). I can't say I've seen this before.

Update: this only happens when we set `JULIA_CPU_TARGET`; if we set it to `generic`, or don't set it, this doesn't happen.  But as soon as we include multiversioning (e.g. just setting `JULIA_CPU_TARGET='generic;sandybridge,-xsaveopt,clone_all'`) it breaks. On `release-1.6`, we can set it to the full typical value of `generic;sandybridge,-xsaveopt,clone_all;haswell,-rdrnd,base(1)` and it doesn't crash.  So something in this backport stack broke it; I will next run a `git bisect` to try and figure out what broke it.

**Comment by KristofferC** [comment]
Thanks for all your work here @staticfloat!

**Comment by staticfloat** [comment]
Bisect points to 93915f8b8d78c401a6e9426a36dac3bd5f88739a.  Manually reverting that on the tip of this branch fixes things. and when I go and look at the original PR, lo and behold, Tim already fixed this in a newer version of LLVM.  So I guess the options are we can either bump LLVM version, or we can drop that backport.

**Comment by KristofferC** [comment]
Let's drop it then.

**Comment by staticfloat** [comment]
The Pkg test failure is that `Base.runtests()` [sets `--depwarn=error`](https://github.com/JuliaLang/julia/blob/backports-release-1.6/test/testenv.jl#L19), and the `release-1.6` Pkg branch has [deprecated `generate`](https://github.com/JuliaLang/Pkg.jl/pull/1923).  Note that [the deprecation was reverted](https://github.com/JuliaLang/Pkg.jl/pull/2492), but it appears that the reversion hasn't been backported to Pkg's `release-1.6` branch.

**Comment by staticfloat** [comment]
I have rebased and dropped the two FP16-related commits.

**Comment by staticfloat** [comment]
@KristofferC can I leave it to you to deal with the Pkg errors?  I think most likely the right thing to do is to [backport the `generate` improvements](https://github.com/JuliaLang/Pkg.jl/pull/2492) to 1.6.

**Comment by KristofferC** [comment]
Of course, will fix it.

**Comment by KristofferC** [comment]
@staticfloat, I think something happened with Float16 now:

```

Test Failed at /cache/build/default-aws-shared0-2/julialang/julia-release-1-dot-6/julia-e6e7dcd667/share/julia/test/numbers.jl:432
--
  | Expression: sprint(show, MIME("text/plain"), Float16(5.0)) == "Float16(5.0)"
  | Evaluated: "Float16(1.526e-5)" == "Float16(5.0)"
`` 

etc.

**Comment by staticfloat** [comment]
@vchuravy can you advise on the Float16 issues here?  I dropped two commits that we backported, since they relied on a newer version of LLVM.

**Comment by simsurace** [comment]
@KristofferC Shouldn't #37299 be included here as well? The docstring that was added in #40486 mentions version 1.6, so isn't this technically a bug?

**Comment by fredrikekre** [comment]
The bug is the docstring, should be 1.7.

**Comment by simsurace** [comment]
> The bug is the docstring, should be 1.7.

Thanks. Opened a PR #48252.

**Comment by PallHaraldsson** [comment]
Simply close this because of new TLS? Or release last point release for some reason (anyone stuck on 1.6.x)?

**Comment by benz0li** [comment]
@KristofferC Please close [to avoid any possible confusion].

**Pull request closed without merging**